### PR TITLE
Add parameter for the starting index for indexOf and lastIndexOf.

### PR DESCRIPTION
**Pull request by JakeWharton**
KT-7067

### CODE REVIEW

File: js/js.libraries/src/core/generated/_ArraysJs.kt [hunk 1]
```diff
@@ -1767,18 +1812,63 @@ public inline fun CharArray.last(predicate: (Char) -> Boolean): Char {
...
+@Deprecated("Use lastIndexOf(T, Int)", level = DeprecationLevel.HIDDEN)
+public fun CharArray.lastIndexOf(element: Char): Int {
+    return lastIndexOf(element, 0)
+}
+
 /**
  * Returns last index of [element], or -1 if the array does not contain element.
  */
-public fun <@kotlin.internal.OnlyInputTypes T> Array<out T>.lastIndexOf(element: T): Int {
+public fun <@kotlin.internal.OnlyInputTypes T> Array<out T>.lastIndexOf(element: T, startIndex: Int = 0): Int {
```
Comment (ilya-g) [inline js/js.libraries/src/core/generated/_ArraysJs.kt]: In string extensions `lastIndexOf` begins search from `startIndex` position (which is `lastIndex` by default) and goes down to zero.

File: libraries/tools/kotlin-stdlib-gen/src/templates/Elements.kt [hunk 2]
```diff
@@ -89,21 +102,33 @@ object Elements : TemplateGroupBase() {
...
     }
 
     val f_lastIndexOf = fn("lastIndexOf(element: T)") {
         include(Iterables, Sequences, ArraysOfObjects, ArraysOfPrimitives, Lists)
+    } builder {
+        typeParam("@kotlin.internal.OnlyInputTypes T")
+        specialFor(Lists) {
+            annotation("""@Suppress("EXTENSION_SHADOWED_BY_MEMBER") // false warning, extension takes precedence in some cases""")
+        }
+        annotation("""@Deprecated("Use lastIndexOf(T, Int)", level = DeprecationLevel.HIDDEN)""")
```
Comment (ilya-g) [inline libraries/tools/kotlin-stdlib-gen/src/templates/Elements.kt]: Use `deprecate(forBinaryCompatibility)` instead.

**Comment by JakeWharton** [comment]
Yeah sorry I've been meaning to get back to this.

**Comment by ilya-g** [comment]
Questions to discuss: 
- should there be overloads with `startIndex` for `Iterable` and `Sequence`, given that they do not have efficient way to skip elements.
- should we leave overloads like `List<T>.indexOf(element: T)`, those that delegate search to the list itself, which may be more efficient.
- how to implement `List.indexOf(element, startIndex)` better:
  - using indexed access
  - using `listIterator(startIndex)`
  - using `subList().indexOf(element)`

**Pull request closed without merging**